### ISSUE TITLE
More flexible Config for underscored attributes, to enable them to be used as regular model fields, like '_id' for MongoDB models.

### ISSUE THREAD

**Issue by AntonOvsyannikov** [samuelcolvin/pydantic#2105]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this feature/change is needed
* [x] After submitting this, I commit to one of:
    * Look through open issues and helped at least one other person
    * Hit the "watch" button on this repo to receive notifications and I commit to help at least 2 people that ask questions in the future
    * Implement a Pull Request for a confirmed bug

# Feature Request

As mentioned in #2101 "ignoring underscore attrs was default behavior for a long time". But now, been having the Config for such attributes it seems quite artificial limitation.

Underscore attributes are considered as private in Python world, but widely used outside with other meanings, i.g. `_id` field for MongoDB documents is a normal field and aliasing it when defining model brings lot of inconvenience.

Pydantic is a library for interacting with the outside world, so suppose it should be more flexible there and provide ability to use underscored attributes as an ordinary regular model field.

Maybe it will be useful to divide underscored attributes to sunder and dunder, but it's discussible and outside of scope of this feature request.

So i propose to introduce the following Config for underscored attrs:
```
underscore_attrs:  Literal['ignored', 'regular', 'private']

# ignored - default behavior
# regular - regular model fields
# private - are private, like underscore_attrs_are_private = True
```

or little bit less descriptive config item

```
underscore_attrs_ignored: boolean = True # by default
```

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.7.2
            pydantic compiled: True
                 install path: /Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pydantic
               python version: 3.7.6 (v3.7.6:43364a7ae0, Dec 18 2019, 14:18:50)  [Clang 6.0 (clang-600.0.57)]
                     platform: Darwin-18.7.0-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions', 'devtools']
```

**Comment by AntonOvsyannikov** [comment]
```
# ignored - default behavior
```

To be honest, can't imagine use case for default behavior of underscored attributes.
They are useless at all - we can nor initialize, nor assign values, and `hasattr` returns False.

```py
from pydantic import BaseModel


class Model(BaseModel):
    _id: int


m = Model(_id=1)  # ignored
print(hasattr(m, '_id'))  # False
m._id = 1  # ValueError: "Model" object has no field "_id"

```

**Comment by flaviovs** [comment]
> Pydantic is a library for interacting with the outside world

Exactly. Note that true private attributes are also affected negatively by how underscore is handled: today, even with `Config.underscore_attrs_are_private = True` one **must** declare all private names as class attributes. This pollutes the attribute list with variables that are not meant to be interacted externally, and it's a PITA while developing since the class private attribute list needs to be maintained and kept in sync with class code.

Since attributes starting with underscore are by convention private, I think the library should ignore them altogether by default, both when assigning and accessing, **except if they're declared as model attributes**. This works nicely with other libraries like MongoDB that use names starting with "_" which, by definition (in the context of those libraries), are public interfaces therefore defining them  as model attributes would make total sense. This also would remove the need of having `Config.underscore_attrs_are_private` and `PrivateAttr` kludges.

**Comment by StephenBrown2** [comment]
There's a reason Object-Oriented libraries for Mongo like MongoEngine use an un-prefixed `id` for the `_id` attribute: https://docs.mongoengine.org/guide/document-instances.html#document-ids

And this Note:
> There is one caveat on Dynamic Documents: fields cannot start with `_`

https://docs.mongoengine.org/guide/defining-documents.html#dynamic-document-schemas

**Comment by justonlyyo** [comment]
if is mongodb model
you can try " ....alias=_xxxx...."
class BaseDoc(BaseModel):
    """
    example
    """
    id: ObjectId = Field(None, alias="_id", title="Primary key", description="mongodb's Primary key")

**Comment by nuno-andre** [comment]
I agree with the _private by default_ behavior, but mangling the sunder attributes once they have been set up as private involves some issues such as losing the type annotations or the need to overload classmethods like `from_orm()` or `construct()`.

And it's true there's a convention of treating sunder attributes as _private_, but it's set on the basis of "we are all consenting adults here". IMHO, the enforcement breaks the PoLA with no significant advantages. Also, it's very common for classes that include some kind of metaprogramming to use sunder public attributes to avoid colliding with user or dynamic attributes, and `BaseModel` falls indeed into this category.

---

@AntonOvsyannikov Private attributes are intended to be managed on initialization. E.g.
```python
from pydantic import BaseModel
from pydantic.fields import Undefined

class Model(BaseModel):
    _private: str  # also PrivateAttr with default value or factory
    attr: str

    class Config:
        underscore_attrs_are_private = True

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        for attr, decl in self.__private_attributes__.items():
            value = kwargs.get(attr, decl.get_default())
            if value is not Undefined:
                setattr(self, attr, value)


obj = Model(attr='ATTR', _private='PRIVATE')
print(obj)
#> attr='ATTR'
print(obj._private)
#> PRIVATE
```

You won't get validation this way, but there's a forthcoming feature that will help with this: #2625. AFAIK, there's not a specific way to do it, and also private attributes lose their annotations, so it would be necessary to use some kludge. E.g.

```python
from pydantic import PrivateAttr

class Model(BaseModel):
    _private: int = PrivateAttr(default_factory=int)

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        for attr, decl in self.__private_attributes__.items():
            value = kwargs.get(attr, decl.get_default())
            if value is not Undefined:
                if decl.default_factory:
                    value = decl.default_factory(value)
                setattr(self, attr, value)

obj = Model(_private='001')
print(obj._private)
#> 1

obj2 = Model()
print(obj2._private)
#> 0
```

Also, you can set the attribute in a slot.
```python
class Model(BaseModel):
    __slots__ = '_private',
    attr: str

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        for attr in self.__slots__:
            try:
                object.__setattr__(self, attr, kwargs[attr])
                # self.__dict__[attr] = getattr(self, attr)
            except KeyError:
                pass
```

Uncommenting the `self.__dict__` line you'll get:
```
print(obj)
#> attr='ATTR' _private='PRIVATE'
```

But this is probably not a good idea for production code.

Cheers!

**Comment by PrettyWood** [comment]
I guess it could be done if we set `Config.underscore_attrs_are_private = True` by default in v2 (see https://github.com/samuelcolvin/pydantic/discussions/2471).
If set to `False` we could accept all names (so make `is_valid_field` more relax) and only refuse dunder names
@samuelcolvin WDYT? Should we add this in v2 milestone?

**Comment by nuno-andre** [comment]
That would be great :) Some ideas FWIW:

- Enable `validate_assignment` for private attributes. Type annotation is currently erased in `PrivateAttr`, could `Annotated(<type>, PrivateAttr)` be a valid approach?

- If the private attributes are not going to be added to `__fields_set__`, passing the kwargs to `_init_private_attributes` would avoid having to subclass the instantiation methods that don't call `__init__` (such as `from_orm` or `construct`).


- Pydantic sets as an invalid field [every attribute that starts with an underscore](https://github.com/samuelcolvin/pydantic/blob/5261fd05a0374b84ce2602d45990baf480fa2417/pydantic/utils.py#L633). Restricting this could be a way to keep both behaviors.

  stdlib's `enum`, for instance, only identifies [attributes surrounded by exactly two underscores (`__name__`) as dunder and by exactly one (`_name_`) as sunder](https://github.com/python/cpython/blob/master/Lib/enum.py#L31). Although maybe the latter is very `enum`-specific and `_name` is a more conventional sunder form, e.g.

  ```python
  import re
  DUNDER = re.compile(r'^__[^_].*?[^_]__$')
  SUNDER = re.compile(r'^_[^_].*?[^_]$')

  assert DUNDER.match('__test_var__')
  assert not DUNDER.match('__test_var___')
  assert SUNDER.match('_test_var')
  assert not SUNDER.match('_test_var_')
  ```

- `Extra` could be a good place to broaden the selection of attributes without modifying the current behavior. For example: `Extra.allow_sunder`, `Extra.allow_dunder`, and `Extra.allow_under` (to allow both).

  Or even accepting a set of names in `Config.extra` that bypasses the `is_valid_field` check (this would also allow specifying a restricted set of optional attributes easily).

  ```python
  class Model(BaseModel):
      class Config:
          extra = {'this', '_and_this', '__and_this__'}
  ```

**Comment by samuelcolvin** [comment]
Happy to allow private attribute fields, but not dunder (`__` prefix) attributes.

This should be behind a flag of some sort with default `False`.

**Comment by samuelcolvin** [comment]
To be clear: I still think using it would almost always be a bad idea, but I don't see any reason why pydantic should prevent it.

**Comment by mwgamble** [comment]
> Underscore attributes are considered as private in Python world, but widely used outside with other meanings, i.g. _id field for MongoDB documents is a normal field and aliasing it when defining model brings lot of inconvenience.

I'm curious why using an alias is inconvenient.

**Comment by samuelcolvin** [comment]
Ye, very good question, I'm not clear why aliases don't solve this, particularly as we now have load and dump aliases.

**Comment by misuzu** [comment]
It surely solves this, but also adds significant cognitive load, just take a look at the things people do to workaround this limitation: https://github.com/tiangolo/fastapi/issues/771 https://github.com/tiangolo/fastapi/issues/1515
Would be nice to just use `_id` field and be done with it.

**Comment by 153957** [comment]
> Ye, very good question, I'm not clear why aliases don't solve this, particularly as we now have load and dump aliases.

For databases which use `_id` using `id` instead and setting up an alias is not always feasible because we have some models which also have an actual `id` field, in which case we have to use something else again for the `_id` field, maybe `id_`.

But as others have said this add a lot of cognitive load, and a lot of extra explaining to out users. With patching of `is_valid_field` this was easily fixed in Pydantic v1, I've not yet found a way to do this in v2. It would be great if this was just supported (perhaps with a config option if necessary).

**Comment by lig** [comment]
@samuelcolvin I'd suggest an annotation similar to `PrivateAttr` but doing the opposite for a single field. Maybe just using an explicit Field.

```py
class Model(BaseModel):
    _private_attr: int
    _id: Annotated[str, Field()]
```
Using `Field` here explicitly makes this attribute a model field despite its name starts with `_`.

**Comment by 153957** [comment]
@samuelcolvin Is this (support for `_` prefixed fields) something that is considered to be supported by pydantic?

**Comment by 153957** [comment]
@Kludex 
As mentioned during our brief discussion yesterday, this issue is currently blocking us from updating to pydantic v2.

We use pydantic for models with ArangoDB. Our base model from which all others are inherited looks something like this:
```python
class ArangoModel(BaseModel):
    class Config:
        extra = 'forbid'
        use_enum_values = True
        validate_assignment = True

    _id: str = None
    _key: str = None
    _rev: str = None
```
And for edge models:
```python
class Edge(ArangoModel):
    _from: str = None
    _to: str = None
```
Other models for specific vertex/edge collections inherit from these and specify additional fields.

There are many cases where access these base fields on initialized models, and in some cases specifically set them:
```python
other_model = SpecificModel(**returned_dict_from_database_interface)
model._id = 'table_name/some_unique_id'
edge._from = model._id
edge._to = other_model._id
```

And we want these to be included using the same names (`_id`) when converting them to a dict to send these to the database, and when retrieving objects from the database.
Currently we use the `python-arango` package which already handles (de)serialization from json to python and back, perhaps when using v2 we can try to make pydantic create models directly from the json.

So we hope pydantic can support (single-)underscore prefixed fields and treat them as regular fields for assignment, serialization, deserialization, etc.. I understand this might not be the default, so it would be fine if this has to be configured via the model `ConfigDict`.
Thanks for considering this.

**Comment by Opa-** [comment]
Here's a workaround if you're in the same context as me :
- You want MongoDB to handle `_id` generation
- You only need the `id` field in a response body and not in a request
- You cannot use `response_model_by_alias` because you need the API to be camelCase with `alias_generator = to_camel`

You just need to override Pydantic constructor in order to rename the `_id` key to `id` in the incoming data.

```py

# Represents an ObjectId field in the database.
# It will be represented as a `str` on the model so that it can be serialized to JSON.
PyObjectId = Annotated[str, BeforeValidator(str)]


class YourObjectBase(BaseModel):
    some_field: str

    class Config:
        alias_generator = to_camel
        populate_by_name = True


class YourObjectResponse(YourObjectBase):
    id: PyObjectId

    """
    Overriding Pydantic constructor because of some issues between Pydantic and MongoDB.
    MongoDB uses _id as identifier while Pydantic consider fields beginning with _ as private.
    We cannot use Pydantic alias either because it would generate an incorrect schema documentation.
    
    For more information, see :
        - https://github.com/tiangolo/fastapi/issues/1515
        - https://github.com/pydantic/pydantic/issues/2105
    """
    def __init__(__pydantic_self__, **data: Any) -> None:
        if '_id' in data:
            data['id'] = data.pop('_id')
        super().__init__(**data)
```

**Comment by reith** [comment]
My code deserializes requests coming from another service, ClientService, as `MessagePayload` model. ClientService, will set `previous_run_id` in future but today it doesn't. ClientService, luckily, sends information I can derive `previous_run_id` from. I think this is natural to define `MessagePayload` like:

```python
from pydantic import BaseModel

_unset_placeholder = ""

class MessagePayload(BaseModel):
    previous_run_url: str
    # NOTE: previous_run_id will be set by ClientService in future
    _previous_run_id: str | None = Field(_unset_placeholder, alias="previous_run_id")"

    @property
    def calculated_previous_run_id(self) -> str | None:
        if self._previous_run_id != _unset_placeholder:
            return self._previous_run_id

        return derive_previous_run_id_from_url(self.previous_run_url)
```

I like `_previous_run_id` be private in my model definition so that I could easily catch if I mistakenly used that field, instead of `calculated_previous_run_id` in my code. Of course, when I learn ClientService has changed to always supply `previous_run_id`, I will make that field _public_ and drop `calculated_previous_run_id`.

**Comment by 153957** [comment]
@sydney-runkle Could support for underscore prefixed field names be considered for Pydantic v3? (if it would be considered to big of a change to support in v2)
(And for those fields to behave the same as any non-underscore prefixed field.)
